robottelo/hosts.py

```diff
@@ -479,6 +480,119 @@ def install_cockpit(self):
+        :param org: Organization to register content host for, required.
+        :param loc: Location to register content host for, required.
+        :param activation_keys: Activation key name to register content host
+            with, required.
+        :param setup_insights: Install and register Insights client, requires OS repo.
+        :param setup_remote_execution: Copy remote execution SSH key.
+        :param setup_remote_execution_pull: Deploy pull provider client on host
+        :param lifecycle_environment: Lifecycle environment.
+        :param operating_system: Operating system.
+        :param packages: A list of packages to install on the host when registered.
+        :param repo: Repository to be added before the registration is performed, supply url.
+        :param repo_gpg_key_url: Public key to verify the package signatures, supply url.
+        :param remote_execution_interface: Identifier of the host interface for remote execution.
+        :param update_packages: Update all packages on the host.
+        :param ignore_subman_errors: Ignore subscription manager errors.
+        :param force: Register the content host even if it's already registered.
+        :param insecure: Don't verify server authenticity.
+        :param username: Satellite admin username
+        :param password: Satellite admin password
+        :return: SSHCommandResult instance filled with the result of the
+            registration.
+        """
+        insights = (
+            # requires OS repo enabled for host
+            f'&setup_insights={str(setup_insights).lower()}'
+            if setup_insights is not None
+            else ''
+        )
+        rex = (
+            f'&setup_remote_execution={str(setup_remote_execution).lower()}'
+            if setup_remote_execution is not None
+            else ''
+        )
+        rex_pull = (
+            # requires Satellite Client repo enabled for host
+            f'&setup_remote_execution_pull={str(setup_remote_execution_pull).lower()}'
+            if setup_remote_execution_pull is not None
+            else ''
+        )
+        lce = ''
+        if lifecycle_environment:
+            selected_lce = entities.LifecycleEnvironment().search(
+                query={'search': f'name="{lifecycle_environment}"'}
+            )[0]
+            lce = f'&lifecycle_environment_id={selected_lce.id}'
+        os = ''
+        if operating_system:
+            selected_os = entities.OperatingSystem().search(
+                query={'search': f'name="{operating_system}"'}
+            )[0]
+            os = f'&operating_system_id={selected_os.id}'
+        pkgs = f'&packages={"+".join(packages)}' if packages is not None else ''
+        rex_iface = (
+            f'&remote_execution_interface={remote_execution_interface}'
+            if remote_execution_interface is not None
+            else ''
+        )
+        rp = f'&{urlencode({"repo": repo })}' if repo is not None else ''
```

Comment:
urlencode will do all this formatting for you, it handles the data structure, not just a single field!

```
In [1]: from urllib.parse import urlencode
In [2]: query_params = dict()
In [3]: query_params['update_packages'] = True
In [4]: query_params['organization_id'] = 1
In [5]: query_params['setup_insights'] = True

In [6]: urlencode(query_params)
Out[6]: 'update_packages=True&organization_id=1&setup_insights=True'
```

You just have to deal with whether or not you set a dict key based on if that field should be included or not, but you don't have to try and format every single field individually. Build the dictionary of all the key-value pairs that should be in the query string, and then format that query string with urlencode.

Reply:
I'm not sold on this, in my opinion it would just make the whole thing harder to debug in the future, plus it doesn't reduce the data wrangling. Still I'd need to deal with some sat api quirks myself, e.g.:

```
query_params['packages'] = ['a', 'b', 'c']
urlencode(query_params)
'update_packages=true&organization_id=1&packages=%5B%27a%27%2C+%27b%27%2C+%27c%27%5D'
```
for packages I need `packages=a+b+c`

Reply:
```
In [6]: query_params['packages'] = 'a b c'

In [7]: urlencode(query_params)
Out[7]: 'update_packages=True&setup_insights=True&packages=a+b+c'
```

The default quoting method of urlencode is quote_plus, which will substitute `+` for spaces, you just have to provide the value as a space delineated string.

Reply:
I could modify inputs to the form that urlencode would like, but why? It makes sense to use it for url processing, but for the whole thing I don't see the value. Currently you have at least a notion of what you are creating in that function just by reading the code, you can compare to what you get from satellite, it's just more maintainable. Plus if the default quoting method changes it will create problems in the future

Reply:
We're getting to some subjective arguments here, but I don't think the readability is any better with the string composition you have here than it would be with creating a dictionary that gets encoded at once.  My argument is essentially boiling down to "we should use standard library functions for their intended purpose". 

The value of the urlencoding on the entire thing is just that - every value will be properly encoded. All the fields will be concatenated, quoted, and encoded. You're not hand-composing `f'&something={something}'` strings over and over.  The current code is only encoding *some* of the values, and randomly generated names/ids coming from elsewhere in the framework could include characters that need to be encoded. Can you always be sure lifecycle_environment_id, operating_system_id, remote_execution_interface, and individual package names will never include characters that need to be encoded in a query string?

The current code is even having to explicitly use `'+'.join(packages)` within the string composition to property quote them, while claiming the urlencode pattern is not readable or understandable. `urlencode` will encode the list of packages in the expected way, by default - including encoding any of the package names themselves. You just use `' '.join(packages)` as the value in the dict of all query params, to space separate them instead of passing them as a python list.

What are you comparing to satellite in this case, and how is it any more maintainable?  Using what I'm suggesting, you would have a dictionary with all the key-value query parameters that is very readable when logged, instead of a collection of 15 variables that are all currently not logged or debuggable. You could then log the resulting URL after being encoded. The SSH library will be logging the curl command that gets executed, so you'll see the URL there and be able to directly compare to something from Satellite.  Right now to debug this, you have to log all 15 variables separately, then look at the ssh command log to see the full URL after all the individual variables are formatted in.

The default quoting method won't change, standard library functions don't just change like that.

Reply:
The intent here is to build the url *as Satelllite would do it* with all the quirks and inconsistencies. This can differ form the way urlencode would do it, and more differences can arise in the future. E.g. for multiple AKs sat would do `activation_keys=mykey%2Ctestkey` for multiple packages it would do `packages=one+two`. 

That's why I prefer to explicitly compose those strings, I don't see any harm in that. Going instead with `list input > space separated string > urlencode > + separated string` is just a long way around for the same result. 

Basically, I'd argue that emulating the Satellite's way of building that curl is *not* the intended purpose for urlencode, or at least it doesn't really make things easier. Still you'd have to deal with none states, strange joins and values, etc.  With current impl. you can log the curl command as well + you can read how it's built, non-compliant IDs would be acutally a bug against Sat, you don't want urlencode to take care for them. 

See, we can take this offline or to the separate issue if it's ok with you.